compiledtrees/code_gen.py

```diff
@@ -153,26 +166,27 @@ def code_gen_ensemble(trees, individual_learner_weight, initial_value,
                     weight=individual_learner_weight)
                 gen.write(increment)
             gen.write("return result;")
-    return gen.lines
-
-
-def compile_code_to_object(code):
-    # XXX - should we clean up the temporary directory left from
-    # mkdtemp()?
-    def call(args):
-        DEVNULL = open(os.devnull, 'w')
-        subprocess.check_call(" ".join(args),
-                              shell=True, stdout=DEVNULL, stderr=DEVNULL)
-
-    tmpdir = tempfile.mkdtemp()
-    cpp_f = os.path.join(tmpdir, "tree.cpp")
-    so_f = os.path.join(tmpdir, "tree.so")
-    o_f = os.path.join(tmpdir, "tree.o")
-
-    with open(cpp_f, 'w') as f:
-        f.write(code)
-
-    call([CXX_COMPILER, cpp_f, "-c", "-fPIC", "-o", o_f, "-O3"])
-    call([CXX_COMPILER, "-shared", o_f,
-         "-fPIC", "-flto", "-o", so_f, "-O3"])
+    return tree_files + [gen.file]
+
+def _compile(cpp_f):
+    o_f = tempfile.NamedTemporaryFile(prefix='compiledtrees_', suffix='.o', delete=True)
+    _call([CXX_COMPILER, cpp_f, "-c", "-fPIC", "-o", o_f.name, "-O3"])
+    return o_f
+
+def _call(args):
+    DEVNULL = open(os.devnull, 'w')
+    subprocess.check_call(" ".join(args),
+                          shell=True, stdout=DEVNULL, stderr=DEVNULL)
+
+def compile_code_to_object(files, n_jobs=1):
+    # if ther is a single file then create single element list
+    # unicode for filename; name attribute for file-like objects
+    if type(files) is unicode or hasattr(files, 'name'):
```

Comment:
I have no view on that, whichever you prefer. But, this breaks stuff - if you had only one tree (without an ensemble), then your code will not work anymore when the function expects a list. Anyhow, we'd have to check for single element here, because it fill fail silently (eg. it will iterate over elements of string).